SignalServiceKit/src/Util/TypingIndicators.swift

```diff
@@ -0,0 +1,363 @@
+
+            sendPauseTimer?.invalidate()
+            sendPauseTimer = Timer.weakScheduledTimer(withTimeInterval: 5,
+                                                      target: self,
+                                                      selector: #selector(OutgoingIndicators.sendPauseTimerDidFire),
+                                                      userInfo: nil,
+                                                      repeats: false)
+        }
+
+        func didStopTypeOutgoingInput() {
+            AssertIsOnMainThread()
+
+            // Send ACTION=STOPPED message.
+            sendTypingMessage(forThread: thread, action: .stopped)
+            // Cancel the sendRefresh timer
+            sendRefreshTimer?.invalidate()
+            sendRefreshTimer = nil
+            // Cancel the sendPause timer
+            sendPauseTimer?.invalidate()
+            sendPauseTimer = nil
+        }
+
+        @objc
+        func sendPauseTimerDidFire() {
+            AssertIsOnMainThread()
+
+            // If the sendPause timer fires:
+
+            // Send ACTION=STOPPED message.
+            sendTypingMessage(forThread: thread, action: .stopped)
+            // Cancel the sendRefresh timer
+            sendRefreshTimer?.invalidate()
+            sendRefreshTimer = nil
+            // Cancel the sendPause timer
+            sendPauseTimer?.invalidate()
+            sendPauseTimer = nil
+        }
+
+        @objc
+        func sendRefreshTimerDidFire() {
+            AssertIsOnMainThread()
+
+            // If the sendRefresh timer fires:
+
+            // Send ACTION=TYPING message
+            sendTypingMessage(forThread: thread, action: .started)
+            // Cancel the sendRefresh timer
+            sendRefreshTimer?.invalidate()
+            // Start the sendRefresh timer for 10 seconds again
+            sendRefreshTimer = Timer.weakScheduledTimer(withTimeInterval: 10,
+                                                        target: self,
+                                                        selector: #selector(sendRefreshTimerDidFire),
+                                                        userInfo: nil,
+                                                        repeats: false)
+        }
+
+        func didSendOutgoingMessage() {
+            AssertIsOnMainThread()
+
+            // If the user sends the message:
+
+            // Cancel the sendRefresh timer
+            sendRefreshTimer?.invalidate()
+            sendRefreshTimer = nil
+            // Cancel the sendPause timer
+            sendPauseTimer?.invalidate()
+            sendPauseTimer = nil
+        }
+
+        private func sendTypingMessage(forThread thread: TSThread, action: TypingIndicatorAction) {
+            Logger.verbose("\(TypingIndicatorMessage.string(forTypingIndicatorAction: action))")
+
+            let message = TypingIndicatorMessage(thread: thread, action: action)
+            messageSender.sendPromise(message: message)
+                .done {
```

Comment:
This logging is redundant with what's already in MessageSender isn't it?

This branch isn't based on master yet, so doesn't have the durable send stuff, but I assume we want non-durable sends for this stuff anyway.

Reply:
Logging removed.  There's a bunch of temporary logging in this class that I intend to remove once the feature is basically working.

Reply:
> This branch isn't based on master yet, so doesn't have the durable send stuff, but I assume we want non-durable sends for this stuff anyway.

Right, these messages are ephemeral.